### PR TITLE
Bubblegum tests first iteration

### PR DESCRIPTION
**Pull request by alexandruag**
This PR presents an initial iteration of the code used to test Bubblegum, together with simple happy-path tests for several operations. We'll subsequently add support for all operations, as well as more scenarios together with testing operations at higher volume to cover more corner cases.

One significant change compared to other programs in MPL is that tests in this PR have been written as unit tests (vs integration tests for other contracts). More details about the distinction between the two categories are available [here](https://doc.rust-lang.org/book/ch11-03-test-organization.html). Unit tests provide more flexibility when needed, as well as making it easier to work with reusable boilerplate. At the end of the day, if we decide to keep all testing done through integration tests, the conversion is straightforward.

### CODE REVIEW

File: bubblegum/program/src/tests/simple.rs [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+use solana_sdk::signature::{Keypair, Signer};
+use solana_sdk::system_instruction;
+use solana_sdk::transaction::Transaction;
+
+use crate::state::metaplex_adapter::{Creator, MetadataArgs, TokenProgramVersion};
+use crate::tests::clone_keypair;
+
+use super::{program_test, Error, LeafArgs, Result, Tree};
+
+async fn fund_account(ctx: &mut ProgramTestContext, address: Pubkey) -> Result<()> {
```
Comment (danenbm) [inline bubblegum/program/src/tests/simple.rs:12]: Is there an `airdrop` functionality you could use?

**Review by danenbm — CHANGES_REQUESTED** [review]
Overall these abstractions look helpful to me and I see them in action in the simple test case and it looks good to me.

It does kind of make it so that you learn the new interface rather than the contract interface, so maybe it could be argued that it is adding a layer of abstraction/complexity to the testing that could have its own opportunity for quirks/bugs/maintenence and takes away from the simplicity of seeing the contract calls in action directly.

But to me it seems fine and the helper objects will reduce copy and paste, and that seems better even if it adds a bit of new layer.   And the new things such as `tx_builder` seem like good patterns with straightforward usage.

File: bubblegum/program/src/tests/simple.rs [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+}
+
+// Test for multiple combinations?
+const MAX_DEPTH: usize = 20;
+const MAX_BUF_SIZE: usize = 64;
+
+// TODO: test signer conditions on mint_authority and other stuff that's manually checked
+// and not by anchor (what else is there?)
+#[tokio::test]
+async fn test_simple() {
```
[the remark this answers is not in the document]
Comment (alexandruag) [reply bubblegum/program/src/tests/simple.rs:43]: done reorganizing tests

File: bubblegum/program/src/tests/mod.rs [hunk 3]
```diff
@@ -0,0 +1,712 @@
...
+use solana_sdk::account::Account;
+use solana_sdk::signature::{Keypair, Signer, SignerError};
+use solana_sdk::signer::signers::Signers;
+use solana_sdk::transaction::Transaction;
+use spl_compression::state::ConcurrentMerkleTreeHeader;
+use spl_concurrent_merkle_tree::concurrent_merkle_tree::ConcurrentMerkleTree;
+
+use crate::state::metaplex_adapter::MetadataArgs;
+use crate::state::request::MintRequest;
+use crate::state::TreeConfig;
```
[the remark this answers is not in the document]
Comment (danenbm) [reply bubblegum/program/src/tests/mod.rs]: Why is it jarring?  I actually find the combined ones easier to read because the whitespace and nesting makes the relationships clearer.  But this is probably an unsolvable debate where half of people like one way and half of people like the other way.

[the remark this answers is not in the document]
Comment (alexandruag) [reply bubblegum/program/src/tests/mod.rs]: Sometimes, when multiple levels get aggregated on single line, it becomes much more difficult to scan for a particular item when you want to leverage stuff like the `use`s being sorted; however I don't feel super strongly either way and I'll gladly conform to the convention currently in use. Also, I think I missed unifying one or two paths, will fix.

File: bubblegum/program/src/tests/simple.rs [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+}
+
+#[tokio::test]
+async fn test_delegate_passes() {
+    let (_, tree, mut leaf) = context_tree_and_leaf().await.unwrap();
+    let new_delegate = Keypair::new();
+
+    tree.delegate(&leaf, new_delegate.pubkey()).await.unwrap();
+    // Reflect changes.
+    leaf.delegate = new_delegate;
```
Comment (danenbm) [inline bubblegum/program/src/tests/simple.rs:64]: So for tests like this where you don't use the local state again, is this update to reflect the change just being done for posterity like if someone were to use this as a basis for a more advanced test, or add more things afterwards?  My assumption is that `leaf` is dropped at the end of this test.

**Review by blockiosaurus — APPROVED** [review]

File: bubblegum/program/src/tests/simple.rs (hunk 4 above)
Comment (alexandruag) [reply bubblegum/program/src/tests/simple.rs:64]: Yes, within the scope of this PR it's for illustrative purposes, but there are some more testing checks coming afterwards where we're actually attempting operations that use/need the new values.

File: bubblegum/program/src/tests/mod.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (danenbm) [reply bubblegum/program/src/tests/mod.rs]: ok I see, sounds good, thanks!

File: bubblegum/program/src/tests/mod.rs [hunk 5]
```diff
@@ -0,0 +1,712 @@
...
+
+#[derive(Debug)]
+pub enum Error {
+    AccountNotFound(Pubkey),
+    Anchor(anchor_lang::error::Error),
+    BanksClient(BanksClientError),
+    Pod(PodCastError),
```
[the remark this answers is not in the document]
Comment (danenbm) [reply bubblegum/program/src/tests/mod.rs]: ok thanks!

File: bubblegum/program/src/tests/simple.rs (hunk 1 above)
Comment (danenbm) [reply bubblegum/program/src/tests/simple.rs]: ok

File: bubblegum/program/src/tests/simple.rs (hunk 4 above)
Comment (danenbm) [reply bubblegum/program/src/tests/simple.rs:64]: ok

**Review by danenbm — APPROVED** [review]
LGTM, you will probably need to rebase on master though there's been some changes such as account name changes from my other PR.

**Review by danenbm — DISMISSED** [review]
LGTM!

**Review by blockiosaurus — APPROVED** [review]

**Review by danenbm — APPROVED** [review]
